handlers/soa_resource.go

```diff
@@ -0,0 +1,121 @@
+package handlers
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/companieshouse/chs.go/log"
+	"github.com/companieshouse/insolvency-api/dao"
+	"github.com/companieshouse/insolvency-api/models"
+	"github.com/companieshouse/insolvency-api/service"
+	"github.com/companieshouse/insolvency-api/transformers"
+	"github.com/companieshouse/insolvency-api/utils"
+	"github.com/gorilla/mux"
+)
+
+// HandleCreateStatementOfAffairs receives a statement of affairs to be stored against the Insolvency case
+func HandleCreateStatementOfAffairs(svc dao.Service) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		vars := mux.Vars(req)
+		transactionID := utils.GetTransactionIDFromVars(vars)
+		if transactionID == "" {
+			log.ErrorR(req, fmt.Errorf("there is no transaction ID in the URL path"))
+			m := models.NewMessageResponse("transaction ID is not in the URL path")
+			utils.WriteJSONWithStatus(w, req, m, http.StatusBadRequest)
+			return
+		}
+
+		log.InfoR(req, fmt.Sprintf("start POST request for submit statement of affairs with transaction id: %s", transactionID))
+
+		// Check if transaction is closed
+		isTransactionClosed, err, httpStatus := service.CheckIfTransactionClosed(transactionID, req)
+		if err != nil {
+			log.ErrorR(req, fmt.Errorf("error checking transaction status for [%v]: [%s]", transactionID, err))
+			m := models.NewMessageResponse(fmt.Sprintf("error checking transaction status for [%v]: [%s]", transactionID, err))
+			utils.WriteJSONWithStatus(w, req, m, httpStatus)
+			return
+		}
+		if isTransactionClosed {
+			log.ErrorR(req, fmt.Errorf("transaction [%v] is already closed and cannot be updated", transactionID))
+			m := models.NewMessageResponse(fmt.Sprintf("transaction [%v] is already closed and cannot be updated", transactionID))
+			utils.WriteJSONWithStatus(w, req, m, httpStatus)
+			return
+		}
+
+		var request models.StatementOfAffairs
+		err = json.NewDecoder(req.Body).Decode(&request)
+
+		// Request body failed to get decoded
+		if err != nil {
+			log.ErrorR(req, fmt.Errorf("invalid request"))
+			m := models.NewMessageResponse(fmt.Sprintf("failed to read request body for transaction %s", transactionID))
+			utils.WriteJSONWithStatus(w, req, m, http.StatusBadRequest)
```

Comment:
If there was an error decoding the request body into our struct, then I think we should be returning an `Internal Server Error` instead of a Bad Request